textsplitter/split_documents.go

```diff
@@ -8,33 +8,41 @@ import (
 	"github.com/tmc/langchaingo/schema"
 )
 
-// ErrMismatchMetadatasAndText is returned when the number of texts and metadatas
+// ErrMismatchMetadatasAndTextAndCustomIDs is returned when the number of texts and metadatas
 // given to CreateDocuments does not match. The function will not error if the
 // length of the metadatas slice is zero.
-var ErrMismatchMetadatasAndText = errors.New("number of texts and metadatas does not match")
+var ErrMismatchMetadatasAndTextAndCustomIDs = errors.New("number of texts, metadatas and custom IDs do not match")
 
 // SplitDocuments splits documents using a textsplitter.
 func SplitDocuments(textSplitter TextSplitter, documents []schema.Document) ([]schema.Document, error) {
 	texts := make([]string, 0)
 	metadatas := make([]map[string]any, 0)
+	customIDs := make([]*string, 0)
 	for _, document := range documents {
 		texts = append(texts, document.PageContent)
 		metadatas = append(metadatas, document.Metadata)
+		customIDs = append(customIDs, document.CustomID)
 	}
 
-	return CreateDocuments(textSplitter, texts, metadatas)
+	return CreateDocuments(textSplitter, texts, metadatas, customIDs)
 }
 
 // CreateDocuments creates documents from texts and metadatas with a text splitter. If
 // the length of the metadatas is zero, the result documents will contain no metadata.
 // Otherwise, the numbers of texts and metadatas must match.
-func CreateDocuments(textSplitter TextSplitter, texts []string, metadatas []map[string]any) ([]schema.Document, error) {
+func CreateDocuments(
+	textSplitter TextSplitter, texts []string, metadatas []map[string]any, customIDs []*string,
```

Comment:
I don't love adding to the palette list like this. I think we could come up with a cleaner interface here. @eliben curious if you'd like to weigh in here?

Reply:
It'd be good to start with a clear explanation of what the PR is trying to accomplish - it seems to be touching much more than just PGVector. Some sort of high-level design would be nice (could be just explained in a PR comment / or in a code comment)

Reply:
The idea behind the `custom_id` field in the PGVector implementation is that you can use to easily relate back to the source of the embedding. In my case for example, the embeddings I insert are coming from a document list in a different database. Being able to specify the custom ID allows me to accomplish a couple of things:

- I can now easily check if embeddings for a database record in the other database exists already
- It makes it easier to selectively delete embeddings that belong together
- When you do a similarity search, you can use the custom ID to relate it back to the original source

When using PGVector, you could probably do the same by storing that ID in the document metadata, but I'm not entirely sure how well that translates to other vector databases. Having it as a separate column also makes it easier to ensure a database index is present to speed up the searches.

The initial idea was to use a single custom ID for each call to `AddDocuments` on the vector store. However, I realised that this approach was too limited and that it would be better to have a custom ID for each individual document that needs to be added to the vector store.

The reason I added it to the Pinecone implementation as well is because I saw that #448 was trying to fix the exact same problem.

Without the option to specify the custom_id for each embedding, that field is kind of useless in the database. Before this PR, a random UUID would get inserted. This serves no purpose. Each record already has a unique UUID in the `uuid` field anyway. Storing two UUIDs for each embedding with one having no proper purpose is just a waste of database resources, right? If you don't want to use a custom ID, we should just use a null-value in the database.

I'll create an new pull request based on this one with a proposition for a cleaned up interface for the `SplitDocuments` and `CreateDocuments` functions.

Reply:
Thanks for the clarification - I'm still not sure I understand how all the pieces of this PR relate together, so I'm leaving more specific questions.

PS: The examples should all build now with the latest alpha tag